[Spring 체스 - 2단계] 스컬(김윤기) 미션 제출합니다.

skullkim opened the pull request:
안녕하세요! 루피! 구현은 모두 완료했어요. 그런데 테스트 코드에서 문제가 하나 있어요.
ChessControllerTest.java에서 EnterGameTest와 DeleteGameTest는 개별적으로 실행하면 통과되지만 기타 테스트 코드와 함께 돌리면 통과가 안되요... 이유를 찾기 위해 디버깅을 했더니 같이 실행 할때만 테스트의 대상이 되는 매핑된 메서드를 찾지 못해 그대로 404가 뜨는 것을 발견했어요. 혼자 수정 해보려 했지만 도저히 뭐가 문제인지 모르겠어서 그대로 제출해요. 혹시 힌트라고 주실 수 있나요?

Ohzzi committed:
feat: 게임 방 번호를 입력하면 게임으로 입장되도록 URL 매핑 구현

Ohzzi committed:
feat: 게임 정보를 불러오는 api 기능과 URL 매핑 완료

Ohzzi committed:
chore: 불필요한 라이브러리 의존성 제거

Ohzzi committed:
refactor: GameDao가 JdbcTemplate을 사용하도록 변경

Ohzzi committed:
refactor: @Transactional 어노테이션을 추가하여 deleteAll 메서드를 대체

Ohzzi committed:
fix: 데이터베이스 설정 수정

Ohzzi committed:
refactor: DAO가 DataSource만 주입받도록 변경

Ohzzi committed:
refactor: PieceDao가 JdbcTemplate을 사용하도록 변경

Ohzzi committed:
refactor: Service를 Spring의 Service로 변경

Ohzzi committed:
refactor: 게임을 시작하는 api 기능을 Spring Controller로 이식

Ohzzi committed:
refactor: 게임을 재시작하는 api 기능을 Spring Controller로 이식

Ohzzi committed:
refactor: 체스 기물을 움직이는 api 기능을 Spring Controller로 이식

Ohzzi committed:
refactor: 점수를 출력하는 api 기능을 Spring Controller로 이식

Ohzzi committed:
refactor: 게임을 종료하는 api 기능을 Spring Controller로 이식

Ohzzi committed:
refactor: 예외 발생 시 예외 메시지를 클라이언트로 보내주는 기능을 Spring Controller로 이식

Ohzzi committed:
chore: 불필요한 파일 및 의존성 제거

Ohzzi committed:
refactor: 체스 보드를 만드는 메서드 분리

Ohzzi committed:
refactor: 게임 조회와 생성 로직 분리

Ohzzi committed:
refactor: 게임 시작 및 재시작 로직을 한 메서드와 URL로 통합

Ohzzi committed:
refactor: 체스 기물 이동 URL 및 메서드 변경

Ohzzi committed:
refactor: 게임 상태 조회 기능 URL 변경

Ohzzi committed:
refactor: 게임 조회 시 게임이 없으면 404 NOT FOUND 반환하도록 변경

Ohzzi committed:
refactor: api URL 간소화

Ohzzi committed:
refactor: Service에서 안쓰는 메서드 제거

skullkim committed:
feat(index.html): 프론트 비밀변호 입력 기능 완료

프론트에서 방에 입장 또는 방을 생성할 떄 비밀버호를 입력하는 기능 완료

skullkim committed:
refactor(index.html): 게임 생성을 위해 입력하는 데이터 이름 변경

게임 생성하기 위해 입력해야 하는 데이터를 게임 id와 비밀번호에서 게임
이름과 비밀번호로 변경

as-is:
 <form method="get" id="game-id-form" action="/game">
  <input type="text" name="gameId" />
    ...
</form>

to-be:
 <form method="get" id="game-id-form" action="/game">
  <input type="text" name="gameId" />
    ...
</form>

skullkim committed:
refactor(init.sql): 테이블 컬럼 변경

게임 방에 비밀번호 적용을 위해 테이블 컬럼 변경

BREAKING CHANGE: game 테이블에 name과 password 컬럼 변경
as-is:
CREATE TABLE game
(
    game_id int         NOT NULL PRIMARY KEY,
    state   varchar(20) NOT NULL
);

CREATE TABLE piece
(
    piece_id   int         NOT NULL AUTO_INCREMENT PRIMARY KEY,
    game_id    INT         NOT NULL,
    position   VARCHAR(10) NOT NULL,
    piece_type VARCHAR(10) NOT NULL,
    color      VARCHAR(10) NOT NULL,
    FOREIGN KEY (game_id)
        REFERENCES game (game_id) ON DELETE CASCADE
);

to-be:
CREATE TABLE game (
    id INT AUTO_INCREMENT NOT NULL,
    name VARCHAR(20) NOT NULL,
    password VARCHAR(20) NOT NULL,
    state VARCHAR(20) NOT NULL,
    PRIMARY KEY (id)
);

CREATE TABLE piece (
    id INT AUTO_INCREMENT NOT NULL,
    position  VARCHAR(10) NOT NULL,
    piece_type VARCHAR(10) NOT NULL,
    color VARCHAR(10) NOT NULL,
    game_id INT NOT NULL,
    PRIMARY KEY (id),
    FOREIGN KEY (game_id) REFERENCES game (id) ON DELETE CASCADE
);

skullkim committed:
refactor(GameDaoImpl.java): 테이블 조작 로직 변경

게임 로직에 비밀번호를 적용하기 위해 테이블을 조작하는 로직 변경

BREAKING CHANGE:
as-is:
@Repository
public class GameDaoImpl implements GameDao {

   ...

    @Override
    public void save(long id) {
        Map<String, Object> parameters = new HashMap<>();
        parameters.put("game_id", id);
        parameters.put("state", GameState.READY);
        insertGame.execute(parameters);
    }

    @Override
    public Optional<GameState> load(long id) {
        String sql = "SELECT state FROM game WHERE game_id = :game_id";

        SqlParameterSource namedParameters = new MapSqlParameterSource("game_id", id);

        try {
            return Optional.ofNullable(
                    namedParameterJdbcTemplate.queryForObject(sql, namedParameters, GameState.class));
        } catch (EmptyResultDataAccessException e) {
            return Optional.empty();
        }
    }

    @Override
    public void updateState(long id, GameState gameState) {
        String query = "UPDATE game SET state = :game_state WHERE game_id = :game_id";
        ...
    }

    @Override
    public void delete(long id) {
        String query = "DELETE FROM game WHERE game_id = :game_id";
        ...
    }
}

to-be:
@Repository
public class GameDaoImpl implements GameDao {

   ...

    @Override
    public void save(String name, String password) {
        Map<String, Object> parameters = new HashMap<>();
        parameters.put("name", name);
        parameters.put("password", password);
        parameters.put("state", GameState.READY);
        insertGame.execute(parameters);
    }

    @Override
    public Optional<GameDto> load(String name, String password) {
        String sql = "SELECT id, state FROM game WHERE name = :name AND password = :password";
        SqlParameterSource namedParameters =
                new MapSqlParameterSource(Map.of("name", name, "password", password));

        try {
            return Optional.ofNullable(
                    namedParameterJdbcTemplate.queryForObject(sql, namedParameters, (res, rowNumber) ->
                            new GameDto(
                                    res.getInt("id"),
                                    GameState.valueOf(res.getString("state"))
                            )));
        } catch (EmptyResultDataAccessException e) {
            return Optional.empty();
        }
    }

    @Override
    public void updateState(long id, GameState gameState) {
        String query = "UPDATE game SET state = :game_state WHERE id = :game_id";
        ...
    }

    @Override
    public void delete(long id) {
        String query = "DELETE FROM game WHERE id = :game_id";
        ...
    }
}

skullkim committed:
refactor(PieceDaoTest.java): 테스트 코드 수정

game 테이블 컬럼 변경에 따른 테스트 코드 수정

BREAKING CHANGE: 테스트 코드의 @BeforeEach 부분 수정

as-is:
@SpringBootTest
@Transactional
public class PieceDaoTest {
    ...

    @BeforeEach
    void createGame() {
        gameDao.save(testGameId);
    }

    ...
}

to-be:
@SpringBootTest
@Transactional
public class PieceDaoTest {
    ...

    @BeforeEach
    void createGame() {
        gameDao.save("name", "password");
        testGameId = gameDao.load("name", "password")
               .get().getId();
    }

    ...
}

skullkim committed:
feat(chess/): 비밀번호 추가 완료

게임 구동시 방 입장을 위한 비밀번호 추가 완료

skullkim committed:
refactor(dao/): 필요없는 인터페이스 삭제

더블 테스트를 하기 위해 생성했던 인터페이스 삭제

BREAKING CHNAGE: 각 DAO에 대응되는 인터페이스 삭제
as-is:
@Repository
public class GameDaoImpl implements GameDao {
    ...
}

@Repository
public class PieceDaoImpl implements PieceDao {
    ...
}

to-be:
@Repository
public class GameDao {
    ...
}

@Repository
public class PieceDao {
    ...
}

skullkim committed:
feat(static/): 모달창 완료

방 리스트를 보여주기 위한 모달창 기능 구현 완료

skullkim committed:
feat(static/): 룸 표시 디자인 완료

게임 룸 하나에 대한 정보를 표시하는 부분 디자인 완료

skullkim committed:
feat(chess/): 현재 존재하는 개임 방 조회 기능 완료(백엔드)

현재 존재하는 게임방 리스트를 조회하고 응답하는 기능 완료

skullkim committed:
feat(static/): 방 조회 기능 완료

프론트에서 생성된 방들의 이름을 조회할 수 있는 기능 완료

skullkim committed:
feat(chess/): 비밀번호 매칭 기능 완료(백엔드)

기존에 생성된 방에 입장하기 전 입력한 비빌번호 일치 여부를 확인하는 기능
완료

skullkim committed:
feat(index.js): 방 입장 기능 완료

게임 방 리스트에서 방 이름을 클릭하고 비밀번호를 입력해 방으로 입장하는
기능 완료

skullkim committed:
refactor(index.js): 게임 입장 로직 분리

기존 방 번호를 선택하고 비밀번호를 입력해서 방으로 입장하는 로직을
별도의 함수로 분리

BREAKING CHANGE: 비밀번호를 입력해 해당 방으로 입장하는 로직을 enterGame
함수로 분리
as-is:
...
document.addEventListener('click', async ({target: {className, id, innerText}}) => {
  if (className !== 'game__title') {
    return;
  }

  const password = prompt("Please input chess game password.");
  const requestJson = JSON.stringify({
    "gameId": id,
    password
  });

  const res = await fetch(`/games/existed-game/${id}`, {
    method: "POST",
    headers: {
      'content-type': 'application/json'
    },
    body: requestJson
  });
  if (res.status === 204) {
    location.href = `/game/${id}`;
    return;
  }
  const {message} = await res.json();
  alert(message);
});

to-be:
...

document.addEventListener('click', async ({target: {className, id}}) => {
  if (className === 'game__title') {
    await enterGame(id);
  }
});

async function enterGame(gameId) {
  const password = inputPassword();
  const requestJson = JSON.stringify({
    "gameId": gameId,
    password
  });

  const res = await fetch(`/games/existed-game/${gameId}`, {
    method: "POST",
    headers: {
      'content-type': 'application/json'
    },
    body: requestJson
  });
  if (res.status === 204) {
    location.href = `/game/${gameId}`;
    return;
  }
  const {message} = await res.json();
  alert(message);
}

function inputPassword() {
  return prompt("Please input chess game password.");
}

skullkim committed:
feat(index.js): 게임 지우는 로직 완료

게임이 끝난 상태이고 올바른 비밀번호를 입력하면 게임을 지우는 로직 완료

skullkim committed:
chore(/): 필요없는 코드 삭제

skullkim committed:
chore(index.js): 필요 없는 코드 삭제

skullkim committed:
refactor(GameAccessRequest.java): 필드 타입 변경

타입을 일치시키기 위해 아이디를 저장하는 필드 타입을 long으로 변경

BREKAING CHANGE: gameId 필드 타입을 int에서 long으로 변경
as-is:
public class GameAccessRequest {

    private int gameId;
    ...

    public GameAccessRequest() {

    }

    public GameAccessRequest(int gameId, String password) {
        this.gameId = gameId;
        ...
    }

    public int getGameId() {
        ...
    }

    ...
}

public class GameAccessRequest {

    private long gameId;
    ...

    public GameAccessRequest() {

    }

    public GameAccessRequest(long gameId, String password) {
        this.gameId = gameId;
        ...
    }

    public long getGameId() {
        ...
    }

    ...
}

workromancer commented on src/main/java/chess/dao/PieceDao.java:
```diff
@@ -1,20 +1,122 @@
+        return pieceType.createPiece(color);
+    };
 
+<<<<<<< HEAD
```
컨플릭트 해결이 제대로 되지 않은 것 같네요 😂

skullkim replied:
...수정하겠습니다😭

workromancer commented on src/main/java/chess/controller/ChessController.java:
```diff
@@ -58,6 +63,63 @@ public ChessGameResponse endGame(@PathVariable long gameId) {
+        return chessService.findAllGames();
+    }
+
+    @PostMapping("/existed-game/?")
```
? 는 어떤 의미로 사용하신 것인가요 ?

skullkim replied:
'/existed-game/' 뒤에 하나의 경로가 더 들어오는 경우를 매핑하기 위해 사용했습니다. '/existed-game/{gameId}' 처럼 사용해도 되지만 gameId를 파라미터가 아닌 바디에서 받아서 사용하고 있기 때문에 @PathVariable로 받아서 사용할 필요성을 못느껴 '?'를 사용했습니다.

workromancer commented on src/main/java/chess/controller/WebChessController.java:
```diff
@@ -0,0 +1,16 @@
+package chess.controller;
+
+public class WebChessController {
```
사용이 되지 않는 것으로 파악했는데 어떤 의도로 만들어주신 것인가요 ?

skullkim replied:
conflict 해결하면서 삭제했던 파일이 그대로 남아있었네요... 지우겠습니다.

workromancer commented on src/main/java/chess/dao/GameDao.java:
```diff
@@ -1,15 +1,88 @@
 
-public interface GameDao {
+@Repository
+public class GameDao {
```
인터페이스 대신 구현 클래스로 만들어주셨군요 👍

구현체가 하나라면 좋은 선택인 것 같아요

workromancer commented on docker/db/mysql/init/init.sql:
```diff
@@ -1,16 +1,17 @@
+CREATE TABLE game (
+    id INT AUTO_INCREMENT NOT NULL,
+    name VARCHAR(20) NOT NULL,
+    password VARCHAR(20) NOT NULL,
```
`체스 게임을 진행할 수 있는 방을 만들어서 동시에 여러 게임이 가능하도록 하기` 의 경우는 비즈니스 규칙인 것으로 보이는데요.

도메인 객체를 추가하지 않고, 쿼리를 수정하신 이유가 있으실까요 ?!

skullkim replied:
아까 구글밋을 하고 계속 생각을 하면서 도메인과 클린 아키텍처 19장을 요약한 글에 대해 찾아보았어요. 그래도 뭐가 뭔지 모르겠네요... 찾아본 글들도 제대로 이해가 안되요.. 눈가리고 아웅 하는 것처럼 너무 추상적으로 다가와서 당장은 이 질문에 대해 제가 진정으로 생각하는 답을 내놓을 수 없을거 같아요. 그럼에도 지금까지 생각한 내용을 말해 볼게요.

1. '체스 게임을 진행할 수 있는 방을 만들어서 동시에 여러 게임이 가능하도록 하기'는 비즈니스 규칙이 아니다.  
  '체스 게임의 핵심 로직은 체스 게임의 룰이지 체스 게임을 매핑 해주는 것이 아니지 않을까?' 라는 생각을 했어요. 그래고 '매핑을 하는 책임을 가진 객체가 필요할까?' 라는 생각도 했어요. 매핑을 하는 책임을 가지는 객체가 있다고 하면 그 객체는 필드로 게임의 id와 ChessGame을 가지고 있겠죠. 그러고 필드값은 생성자의 인자로 넘기면 되니 id를 인자로 받아 같은 id면 ChessGame을 반환하는 메서드 하나만 가질테고요. 이 행위는 결국에는 Map이랑 같은 역할을 하는데 굳이 하나의 객체로 만들어 주어야 하나? 너무 오버하는거 아닌가? 라는 생각이 드네요. 따라서 도메인으로 만들 필요가 없다는 생각을 했어요.
2. '체스 게임을 진행할 수 있는 방을 만들어서 동시에 여러 게임이 가능하도록 하기'는 비즈니스 규칙이 맞다.
  지금까지 도메인 객체를 만들 때면 상항 요구사항을 중심으로 만들어 왔어요. 요구사항을 보고 주요 명사 라고 생각되는 것들을 각각의 도메인 객체로 만드는 방식이에요. 이 관점에서 2단계 요구사항인 '체스 게임을 진행할 수 있는 방을 만들어서 동시에 여러 게임이 가능하도록 하기'를 보면 '체스 게임'과 '방'이 도메인이 될 수 있을거 같아요. 그러면 '방'을 나타내는 하나의 도메인 객체를 만드는 것이 맞겠죠? 

아까 구글밋에서 말해주신 내용들을 다 이해하지 못한거 같아요. 아직도 뭐가 뭔지 모르겠어요. 위에서 말한 두 개의 의견 중 하나를 그냥 선택해 코드에 반영할 수도 있지만, 스스로도 명확하지 못한 상태에서 코드를 작성하면 그저 미션을 완성하기 위한 핑계가 되는거 같아서 우선은 제 의견을 말하는 정도로만 할게요. 내일 가서 크루들과 코치님들께 좀 더 물어보고, 고민해보고, 조금이라고 명확해 지면 그때 코드에 반영할게요.

skullkim replied:
조금 더 고민을 해보니 게임 룸을 도메인에 추가하는 것이 맞다는 결론이 나왔어요. 도메인은 도메인 자체만으로도 요구사항을 충족시킬 수 있어야 하며, 디비 같은 요소는 하나의 애플리케이션에서 핵심이 아니다 라는 결론을 내리게 되었어요. 그러면서 POJO에 대한 생각도 조금 하게 됬어요. 스프링이 나오기 이전에 EJB가 존재했고, EJB가 너무 복잡해 순수 자바 객체만을 가지고 애플리케이션을 만들려는 시도가 스프링의 시초인걸로 알고 있어요. 이 때, 서비스나 컨트롤러 등으로 사용하기 위한 클래스는 어노테이션을 붙여 스프링 내에서 컨트롤러 또는 서비스로 사용하니 프레임워크에 종속적인, 즉 POJO가 아니다 라는 생각이 드네요. 그러면 스프링에서 말하는 POJO는 결국 도메인 객체들을 말하는 건가요?
도메인이 뭔지, 레이어를 어떻게 나누면 될지에 대한 판단을 하기에는, 저만의 기준을 정하기에는 지식과 경험이 아직 부족한거 같아요. 명확해 질지는 모르겠지만, 그래도 계속 질문을 해보면서 개발을 해야 할거 같네요.

workromancer replied:
고민을 굉장히 많이 해주셨군요…

POJO 에 대한 의견도 굉장히 많이 갈리긴 하는데요. 지금처럼 고민 하시면서 계속해서 코드 작성을 하시다보면 언젠간 스컬도 스컬만의 기준이 생길 날이 올꺼예요.

workromancer reviewed (changes requested):
스컬 안녕하세요~

git 커밋 관련된 이슈부터 해서 여러가지로 고생 많으셨어요. 그러다보니 리뷰도 많이 늦어졌는데요.
그래도 이번 미션 잘 진행해주고 계셔요 ㅎㅎ

관련해서 코멘트 몇 개 남겨두었으니 확인해보시고 리뷰 요청 주세요 !

skullkim committed:
chore(PieceDao.java): conflict 해결

해결되지 않은 conflict 해결

skullkim committed:
chore(chess/): conflict 해결시에 잘못 해결한 부분 수정

skullkim committed:
chore(/test): 잘못 해결한 conflict 수정

테스트 코드에서 제대로 수정하지 않은 conflict 해결

skullkim committed:
fix(ChessController.java): 매핑되는 url 형식 변경

테스트 코드에서 url을 찾지 못해 404 에러가 뜨는 문제를 해결하기 위해 url
형식 변경

BREAKIGN CHANGE: comparPassword, deleteGame url 변경
as-is:
@PostMapping("/existed-game/?")
public ResponseEntity<ErrorResponse> comparePassword(@RequestBody GameAccessRequest game) {
    ...
}

@DeleteMapping("/existed-game/?")
public ResponseEntity<ErrorResponse> deleteGame(@RequestBody GameAccessRequest game) {
    ...
}

to-be:
@PostMapping("/existed-game/{gameId}")
public ResponseEntity<ErrorResponse> comparePassword(@RequestBody GameAccessRequest game) {
    ...
}

@DeleteMapping("/existed-game/{gameId}")
public ResponseEntity<ErrorResponse> deleteGame(@RequestBody GameAccessRequest game) {
    ...
}

skullkim committed:
feat(Room.java): 방 도메인 추가

방 별로 서로 다른 체스 게임을 진행하기 위해 방 도메인 추가

workromancer commented on src/main/java/chess/domain/Room.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
동등성 비교를 어떤 기준으로 할지에 대해서 고민해보셔도 좋을 것 같아요 ㅎㅎ

skullkim replied:
통상적인 상황에서는 모든 필드값을 기준으로 동등성을 비교하게 작성하고 있어요. 하지만 Room 같은 경우는 처음 방에 들어갈 때에는 패스워드와 id를 모두 사용하지만 그 후 부터는 오직 id만을 가지고 방을 식별하고 있어요(그렇지 않으면 매번 비밀번호를 사용자가 입력하게 하거나 비밀번호를 어딘가에 별도로 저장해야 되겠죠). 따라서 Room의 동등성을 비교하는 기준은 id + password, id  두 가지로 나뉠 수 있을거 같아요. 이 때, id + password는 오직 방에 입장할 때, 방을 지울 때 두 가지 케이스에만 사용되고, 로직상으로도 비밀번호를 직접 비교하는 로직이 포함되는 것이 명시적이라 생각했어요. 따라서 Room의 동등성 비교에 id만 넣어 놓았어요. 그래야 게임 입장 후 요청하는 요청에 대해 올바른 처리가 가능하겠죠(Map 컬랙션을 사용하고 있어서요)

workromancer reviewed (approved):
안녕하세요 스컬 !

뭔가 굉장히 많은 이야기를 나눴던 미션인 것 같네요.
정말 고생 많으셨어요. 제가 괜히 너무 많은 고민거리들을 드린 것은 아닌가 걱정이 되긴 했는데, 잘 따라와주셔서 고마워요 ㅎㅎ

이번 미션은 이만 머지하고, 다음 미션에서 지금 하던 고민들을 계속 고민 하시면서 미션을 계속해 나가시죠.
우리에게는 아직 많은 시간이 남아있어서 너무 조급해하지 않으셔도 괜찮을 것 같아요.

이번 미션 고생 많으셨고, 이번 미션은 여기서 이만 머지하도록 하겠습니다 !
